cmd/zoekt-sourcegraph-indexserver/index.go

```diff
@@ -156,30 +196,86 @@ func gitIndex(o *indexArgs, runCmd func(*exec.Cmd) error) error {
 		return err
 	}
 
-	fetchStart := time.Now()
+	var fetchCommits = func(commits []string) error {
+		fetchArgs := []string{"-C", gitDir, "-c", "protocol.version=2", "fetch", "--depth=1", o.CloneURL}
+		fetchArgs = append(fetchArgs, commits...)
+
+		cmd = exec.CommandContext(ctx, "git", fetchArgs...)
+		cmd.Stdin = &bytes.Buffer{}
+
+		return runCmd(cmd)
+	}
 
 	// We shallow fetch each commit specified in zoekt.Branches. This requires
 	// the server to have configured both uploadpack.allowAnySHA1InWant and
 	// uploadpack.allowFilter. (See gitservice.go in the Sourcegraph repository)
-	fetchArgs := []string{"-C", gitDir, "-c", "protocol.version=2", "fetch", "--depth=1", o.CloneURL}
-	var commits []string
+	var branchCommits []string
 	for _, b := range o.Branches {
-		commits = append(commits, b.Version)
+		branchCommits = append(branchCommits, b.Version)
 	}
-	fetchArgs = append(fetchArgs, commits...)
 
-	cmd = exec.CommandContext(ctx, "git", fetchArgs...)
-	cmd.Stdin = &bytes.Buffer{}
+	successfullyFetchedCommitsCount := 0
+	var fetchDuration time.Duration
+
+	allFetchesSucceeded := true
+
+	start := time.Now()
+	err = fetchCommits(branchCommits)
+	fetchDuration += time.Since(start)
+	successfullyFetchedCommitsCount += len(branchCommits)
 
-	err = runCmd(cmd)
-	fetchDuration := time.Since(fetchStart)
 	if err != nil {
-		metricFetchDuration.WithLabelValues("false", repoNameForMetric(o.Name)).Observe(fetchDuration.Seconds())
+		successfullyFetchedCommitsCount -= len(branchCommits)
+		allFetchesSucceeded = false
+
+		metricFetchDuration.WithLabelValues(strconv.FormatBool(allFetchesSucceeded), repoNameForMetric(o.Name)).Observe(fetchDuration.Seconds())
 		return err
 	}
 
-	metricFetchDuration.WithLabelValues("true", repoNameForMetric(o.Name)).Observe(fetchDuration.Seconds())
-	debug.Printf("fetched git data for %q (%d commit(s)) in %s", o.Name, len(commits), fetchDuration)
+	if o.UseDelta {
+		// try fetching prior commits for delta builds
+		// if we're unable to fetch prior commits, we continue anyway
+		// knowing that zoekt-git-index will fall back to a "full" normal build
+
+		repositoryName := buildOptions.RepositoryDescription.Name
+		repositoryID := buildOptions.RepositoryDescription.ID
+
+		existingRepository, err := getRepositoryMetadata(o)
+		if err != nil {
+			return fmt.Errorf("(delta build) failed to get repository metadata: %w", err)
+		}
+
+		if existingRepository == nil {
```

Comment:
I'm not super happy with these nested if statements, but the other solutions that I could think of seemed to clutter the control flow even more.

Reply:
a helper which always returns a list of prior commits will avoid the nested if statements. If the repo doesn't exist on disk, then the list is empty. I don't think we need to log for that case.